bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -112,32 +120,48 @@ impl AuthenticationService {
 
     /// Updates the service to authenticate with the homeserver for the
     /// specified address.
-    pub fn configure_homeserver(&self, server_name: String) -> Result<(), AuthenticationError> {
+    pub fn configure_homeserver(
+        &self,
+        server_name_or_homeserver_url: String,
+    ) -> Result<(), AuthenticationError> {
         let mut builder = Arc::new(ClientBuilder::new()).base_path(self.base_path.clone());
 
-        if server_name.starts_with("http://") || server_name.starts_with("https://") {
-            builder = builder.homeserver_url(server_name)
-        } else {
-            builder = builder.server_name(server_name);
-        }
+        // Remove any URL scheme from the name to attempt discovery first.
+        let server_name = matrix_sdk::sanitize_server_name(&server_name_or_homeserver_url)
+            .map_err(AuthenticationError::from)?;
+
+        builder = builder.server_name(server_name.to_string());
+
+        let client = builder
+            .build()
+            .or_else(|e| {
+                if !server_name_or_homeserver_url.starts_with("http://")
+                    && !server_name_or_homeserver_url.starts_with("http://")
+                {
+                    return Err(e);
+                }
+                // When discovery fails, fallback to the homeserver URL if supplied.
+                let mut builder = Arc::new(ClientBuilder::new()).base_path(self.base_path.clone());
+                builder = builder.homeserver_url(server_name_or_homeserver_url);
+                builder.build()
+            })
+            .map_err(AuthenticationError::from)?;
 
-        let client = builder.build().map_err(AuthenticationError::from)?;
+        let details =
+            RUNTIME.block_on(async /*move*/ { self.details_from_client(&client).await })?;
```

Comment:
The commented `move` obviously need to be removed from here. I left it to ask the question that this was safe to remove from the implementation. Assumed yes given the compile was happy.

Reply:
As long as it compiles, `move` or no `move` hardly matters.